src/ledger-iframe-bridge.ts

```diff
@@ -0,0 +1,312 @@
+  | {
+      action: IFrameMessageAction.LedgerConnectionChange;
+      payload: { connected: boolean };
+    }
+  | ({
+      action: IFrameMessageAction.LedgerMakeApp;
+    } & ({ success: true } | { success: false; error?: unknown }))
+  | {
+      action: IFrameMessageAction.LedgerUpdateTransport;
+      success: boolean;
+    }
+  | ({
+      action: IFrameMessageAction.LedgerUnlock;
+    } & (
+      | { success: true; payload: GetPublicKeyResponse }
+      | { success: false; payload: { error: Error } }
+    ))
+  | ({
+      action: IFrameMessageAction.LedgerSignTransaction;
+    } & (
+      | { success: true; payload: LedgerSignTransactionResponse }
+      | { success: false; payload: { error: Error } }
+    ))
+  | ({
+      action:
+        | IFrameMessageAction.LedgerSignPersonalMessage
+        | IFrameMessageAction.LedgerSignTypedData;
+    } & (
+      | {
+          success: true;
+          payload: LedgerSignMessageResponse | LedgerSignTypedDataResponse;
+        }
+      | { success: false; payload: { error: Error } }
+    ))
+);
+
+type IFrameMessage<TAction extends IFrameMessageAction> = {
+  action: TAction;
+  params?: Readonly<Record<string, unknown>>;
+};
+
+type IFramePostMessage<TAction extends IFrameMessageAction> =
+  IFrameMessage<TAction> & {
+    messageId: number;
+    target: typeof LEDGER_IFRAME_ID;
+  };
+
+export class LedgerIframeBridge implements LedgerBridge {
+  iframe?: HTMLIFrameElement;
+
+  iframeLoaded = false;
+
+  eventListener?: (eventMessage: {
+    origin: string;
+    data: IFrameMessageResponse<IFrameMessageAction>;
+  }) => void;
+
+  isDeviceConnected = false;
+
+  currentMessageId = 0;
+
+  messageCallbacks: Record<
+    number,
+    (response: IFrameMessageResponse<IFrameMessageAction>) => void
+  > = {};
+
+  delayedPromise?: {
+    resolve: (value: boolean) => void;
+    reject: (error: unknown) => void;
+    transportType: string;
+  };
+
+  async init(bridgeUrl: string) {
+    this.#setupIframe(bridgeUrl);
+
+    this.eventListener = this.#eventListener.bind(this, bridgeUrl);
+
+    window.addEventListener('message', this.eventListener);
+  }
+
+  async destroy() {
+    if (this.eventListener) {
+      window.removeEventListener('message', this.eventListener);
+    }
+  }
+
+  async attemptMakeApp(): Promise<boolean> {
+    return new Promise((resolve, reject) => {
+      this.#sendMessage(
+        {
+          action: IFrameMessageAction.LedgerMakeApp,
+        },
+        (response) => {
+          if (response.success) {
+            resolve(true);
+          } else {
+            reject(response.error);
+          }
+        },
+      );
+    });
+  }
+
+  async updateTransportMethod(transportType: string): Promise<boolean> {
+    return new Promise((resolve, reject) => {
+      // If the iframe isn't loaded yet, let's store the desired transportType value and
+      // optimistically return a successful promise
+      if (!this.iframeLoaded) {
+        this.delayedPromise = {
```

Comment:
Why do we need a delayedPromise for this? Couldn't we just persist the desiredTransportType in case the iframe is not loaded?

Reply:
Ah I just noticed that we are just moving it from ledger-keyring.. nevermind then, let's keep this PR as simple as possible. Code improvements can be done in subsequent PRs

Reply:
The delayed promise was needed because all that logic was in the constructor, which cannot be asynchronous.

Now that logic is moved to an asynchronous init method, it can be removed, but that should be done in a subsequent PR.